### PR TITLE
Host irs: support MatmulOp with pre-allocated outputs

### PR DESCRIPTION
**Pull request by samnordmann**
# What
support pre-allocated outputs for matmul

# Why
Prerequesite for https://github.com/NVIDIA/Fuser/pull/3181, where the matmul's output needs to be stored in a slice of a global buffer.

# How 
Use`at::matmul_out` if the output is already associated with a buffer.

### CODE REVIEW

**Comment by samnordmann** [comment]
!build

**Review by nsarka — APPROVED** [review]

File: csrc/host_ir/executor.cpp [hunk 1]
```diff
@@ -281,6 +281,25 @@ void HostIrExecutor::handle(EndCoalescing* end_coalescing) {
...
+void HostIrExecutor::handle(MatmulOp* matmul) {
+  TensorView* a = matmul->inA();
+  TensorView* b = matmul->inB();
+  TensorView* out = matmul->out();
+  NVF_ERROR(
+      expr_evaluator_.isKnown(a) && expr_evaluator_.isKnown(b),
+      "Inputs of the matmul ",
+      matmul->toString(),
+      "must be precomputed before being retrieved");
+  if (expr_evaluator_.isKnown(out)) {
```
Comment (wujingyue) [inline csrc/host_ir/executor.cpp:293]: Thanks -- this is part of #3181 that I'd like to separate out. 

Can you try to address https://github.com/NVIDIA/Fuser/pull/3181#discussion_r1800222396? I love small unit tests with good coverage -- much easier to debug than larger ones.

**Review by wujingyue — COMMENTED** [review]
LGTM otherwise

File: csrc/host_ir/executor.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/host_ir/executor.cpp:293]: We already cover both branches with two separate unit tests. 

- The pre-existing `TEST_F(MatmulHostIrTest, HostIr)` https://github.com/NVIDIA/Fuser/blob/55971f8811243822983edf8510b3fa26524d1f15/tests/cpp/test_host_irs.cpp#L759  tests when we do not pre-allocate the output
- the newly added unit tests with pre-allocated outputs

Comment (wujingyue) [reply csrc/host_ir/executor.cpp:293]: Great!

**Review by wujingyue — APPROVED** [review]

File: tests/cpp/test_host_irs.cpp [hunk 2]
```diff
@@ -790,6 +790,44 @@ TEST_F(MatmulHostIrTest, HostIr) {
...
+  constexpr int64_t K = 128;
+  constexpr int64_t N = 256;
+
+  auto hic = std::make_unique<HostIrContainer>();
+  FusionGuard fg(hic.get());
+
+  TensorView* a = makeContigTensor(3);
+  TensorView* b = makeContigTensor(3);
+  TensorView* c = makeContigTensor(3);
+  auto matmul = IrBuilder::create<MatmulOp>(c, a, b);
```
Comment (wujingyue) [inline tests/cpp/test_host_irs.cpp:805]:
```suggestion
  auto* matmul = IrBuilder::create<MatmulOp>(c, a, b);
```

File: tests/cpp/test_multidevice_overlap.cpp [hunk 3]
```diff
@@ -324,9 +326,8 @@ TEST_F(
   TensorView* tva_j = select(tva, 0, j);
   TensorView* tvc_j = select(tvc, 0, j);
   TensorView* tvc_locally_reduced_j =
-      matmul(tva_j, tvb); // ideally we should use the preallocated global
-                          // buffer tc_locally_reduced, but ExpressionEvaluator
-                          // do not support preallocated output buffer.
+      select(tvc_locally_reduced, 0, stream_index);
+  auto matmul = IrBuilder::create<MatmulOp>(tvc_locally_reduced_j, tva_j, tvb);
```
Comment (wujingyue) [inline tests/cpp/test_multidevice_overlap.cpp:330]:
```suggestion
  auto* matmul = IrBuilder::create<MatmulOp>(tvc_locally_reduced_j, tva_j, tvb);
```

**Review by wujingyue — APPROVED** [review]

**Comment by wujingyue** [comment]
!build

**Pull request merged**